web/app/components-v2/forms/FormModel.ts

```diff
@@ -0,0 +1,266 @@
+type ModifyData<O> = { type: 'modify'; modify: (o: O) => O }
+type Cancel = { type: 'cancel' }
+type EndEdit<O> = { type: 'endEdit'; value: O }
+type Validate = { type: 'validate'; constraint: Constraint }
+type Action<O> = StartEdit | ModifyData<O> | Cancel | EndEdit<O> | Validate
+
+const reducer = <O>(
+  state: InternalFormState<O>,
+  action: Action<O>
+): InternalFormState<O> => {
+  switch (action.type) {
+    case 'modify':
+      const data = action.modify(state.data)
+      return state.editMode
+        ? {
+            ...state,
+            data,
+            hasChanged: state.hasChanged || !deepEqual(state.data, data)
+          }
+        : state
+    case 'startEdit':
+      return {
+        ...state,
+        editMode: true,
+        isSaved: false,
+        hasChanged: false,
+        errors: action.constraint
+          ? validateData(state.data, action.constraint)
+          : []
+      }
+    case 'cancel':
+      return {
+        ...state,
+        editMode: false,
+        data: state.initialData,
+        isSaved: false,
+        errors: []
+      }
+    case 'endEdit':
+      return {
+        ...state,
+        data: action.value,
+        initialData: action.value,
+        editMode: false,
+        isSaved: true,
+        errors: []
+      }
+    case 'validate':
+      const errors = validateData(state.data, action.constraint)
+      return {
+        ...state,
+        errors: deepEqual(errors, state.errors) ? state.errors : errors
+      }
+    default:
+      return state
+  }
+}
+
+export const useForm = <O extends object>(
+  initialState: O,
+  startWithEditMode: boolean = false,
+  constraint?: Constraint | null
+): FormModel<O> => {
+  const user = useUser()
+  const init = useMemo(
+    () =>
+      internalInitialState(
+        initialState,
+        user?.hasWriteAccess ? startWithEditMode : false,
+        constraint
+      ),
+    []
+  )
+
+  const [
+    { data, initialData, editMode, hasChanged, isSaved, errors },
+    dispatch
+  ] = useReducer(reducer, init) as [
+    InternalFormState<O>,
+    Dispatch<ReducerAction<Reducer<O, Action<O>>>>
+  ]
+
+  const globalErrors = useGlobalErrors()
+
+  const startEdit = useCallback(() => {
+    if (user?.hasWriteAccess) {
+      dispatch({ type: 'startEdit', constraint })
+    }
+  }, [user])
+
+  const cancel = useCallback(() => {
+    dispatch({ type: 'cancel' })
+  }, [])
+
+  const updateAt = useCallback(
+    <T>(optic: FormOptic<O, T>, modify: (t: T) => T) => {
+      if (editMode) {
+        dispatch({ type: 'modify', modify: modifyValue(optic)(modify) })
+      }
+    },
+    [editMode]
+  )
+
+  const validate = useCallback(() => {
+    if (constraint && editMode) {
+      dispatch({ type: 'validate', constraint })
+    }
+  }, [constraint, editMode])
+
+  useEffect(() => {
+    validate()
+  }, [validate])
+
+  const save = useCallback(
+    async <T>(
+      api: (data: O) => Promise<ApiResponse<T>>,
+      merge: (data: O, response: T) => O
```

Comment:
Onko mergen palautustyyppi sittenkin `O & T`, vai onko tyyppi `T` tyypin `O` osajoukko?

Reply:
Tyyppi `T` voi olla mitä tahansa. Annetun `merge`-funktion tarkoitus on liittää dataan `O` arbitraatinen paluuarvo `T` (tai palauttaa `O` sellaisenaan mitään muuttamatta, jos paluuarvo on tyhjä tai siinä ei ole mitään `O`:n kannalta kiinnostavaa). Tällöin ei tarvitse rajoittaa itseä vaatimuksella että api palauttaisi aina osajoukon `O`:sta.